### PR TITLE
Screenshots.

### LINKED ISSUE #13: Screenshot support
Since all the infrastructure is already in it might be interesting to consider expanding this library to support output to other formats such as PNG or JPG in the future.

See for example
https://medium.com/@dschnr/using-headless-chrome-as-an-automated-screenshot-tool-4b07dffba79a
https://github.com/schnerd/chrome-headless-screenshots

### PR DESCRIPTION
**Pull request by rhyek**
Related: #13.
This is a work in progress. I need to know if this is the approach you were thinking of.

Basically:
```ts
await hpc.create(html, options) // pdf
await hpc.create(html, options, 'pdf') // pdf
await hpc.create(html, options, 'screenshot') // screenshot
await hpc.createPDF(html, options) // pdf
await hpc.createScreenshot(html, options) // screenshot
```
`CreateOptions` now has:
```ts
  screenshotOptions?: {
    /**
     * The options to pass to Chrome's Page.captureScreenshot.
     *
     * @type {ChromeCaptureScreenshotOptions}
     */
    captureScreenShotOptions?: CaptureScreenshotOptions;

    /**
     * The options to pass to Chrome's Emulation.setDeviceMetricsOverride.
     *
     * @type {DeviceMetricsOverrideOptions}
     */
    deviceMetrics?: DeviceMetricsOverrideOptions;

    /**
     * Wether to capture a full-page screenshot.
     *
     * @type {boolean}
     */
    fullPage?: boolean;
  };
```

### CODE REVIEW

**Comment by rhyek** [comment]
Not sure what is causing those test errors. I'll have a look at it later. @westy92 have you had a chance to take a loot at this?

**Comment by tomasc** [comment]
Thank you @rhyek, very nice! 
I think this is a good approach, I like how the interface matches the PDF one. I assume all the `CompletionTrigger` options would be the same, correct?

**Comment by rhyek** [comment]
They should work the same, yes. Only thing missing should be tests.

**Comment by rhyek** [comment]
Hm, well I was sort of expecting some tests to fail due to timeouts where OCR scanning is involved. The other tests I noticed are basically inconsistencies in OCR scanning vs my environment, which I wasn't expecting. Things like HELLO! being interpreted as HELLO'. Those are easy to fix.
Others are sort of just random timeouts that I've noticed happen occasionally.

Thoughts on these tests?

**Comment by rhyek** [comment]
To be honest, the OCR thing was mostly me having a little fun with writing these tests. I'm honestly more inclined to just remove them altogether instead of figuring these inconsistencies out. IMO, with the PDF tests, screenshots could be assumed to work fine for the most part and a couple of tests should suffice.

**Comment by tomasc** [comment]
Thank you, @rhyek. Perhaps an idea might be to split this PR into two – one for adding the screenshot interface, other for additional OCR option. That way the screenshot feature could be merged and new package released, and the OCR added separately later on. What do you think?

**Comment by rhyek** [comment]
Yeah, not a bad idea. I just wish @westy92 would have share his opinion. Atm I have no idea when or if this might be merged and published.

**Comment by tomasc** [comment]
@rhyek @westy92 usually reacts immediately – I guess he might be on vacation or so.

**Comment by westy92** [comment]
@rhyek Thank you for the contribution!  I plan to take a first look tonight.  Sorry for the delay!

**Comment by rhyek** [comment]
It has been nearly a month since I made this PR. Are there any issues that I can help resolve so this can get merged?

**Comment by rhyek** [comment]
I cleaned up the tests mostly by removing the OCR ones which I feel were unneeded. The failed tests are are an issue of slow processing on the test server.

If there isn't any interest in this PR please close.

**Comment by tomasc** [comment]
@westy92 why is this not being merged please? This would be a welcome addition to the library as far as I am concerned.

**Comment by graingert** [comment]
@rhyek why did you close this?

**Comment by rhyek** [comment]
@graingert, sorry. I interpreted the lack of attention to this PR as the repo owner having either no interest or different plans for this feature, which is fine.

**Comment by graingert** [comment]
@westy92 do you know what's happening with this PR?

**Comment by westy92** [comment]
I still have not had the time to review this in depth.  Since this is a new (large) feature, I plan to investigate methods of generating images, verify this is the best one, and then think about how I want this new functionality integrated into the project.  It will take a lot of time that I have not had yet.  I just reopened this, as I will get to it eventually.

**Comment by westy92** [comment]
If anyone else wishes to review this changeset, please do.  The more eyes, the better.

**Comment by graingert** [comment]
Do you think it might be worth splitting the image generation from the
chrome process reuse via tabs?

**Comment by westy92** [comment]
@graingert I don't see any code here working with tabs, but that implementation should remain separate.

**Comment by rhyek** [comment]
@westy92, he means [this](https://github.com/rhyek/html-pdf-chrome/blob/af7054a4bf80cc6e6a032f9b8d22a6ef800b0a89/src/Generators.ts#L74).

I am already using my fork in production and since multiple PDFs and/or screenshots are being generated concurrently, I have found my application crashes frequently without the tab functionality.

For example, I have a page that displays several (20 or so) screenshots at the same time all generated in real-time with a single chrome instance. Without tabs it does not work.

It is very basic since it is just utilizing `chrome-remote-interface`'s API. IMO, the code for both screenshots and tabs is very simple. I think it would be a good idea to review both at the same time in this PR.

There is already a [test](https://github.com/rhyek/html-pdf-chrome/blob/screenshots/test/index.ts#L357) that utilizes tabs. Currently it only generates 5 concurrent PDFs since the global 15s timeout and slow processing of the test server doesn't allow for more, for some reason. That test itself could use a review as well.

File: .gitignore [hunk 1]
```diff
@@ -38,3 +39,7 @@ jspm_packages
 
 # Transpiled files
 lib
+
+# tesseract.js eng language files
+test/langs/*
+!test/langs/.gitkeep
```
Comment (graingert) [inline .gitignore]: missing trailing new line

**Comment by graingert** [comment]
@rhyek do you have a fork on npm? and can you separate out the use of tabs for concurrency?

**Comment by rhyek** [comment]
I'm using my fork from Github, directly. Changing this branch would not
affect me at all. Nevertheless, are your plans for tabs much different than
what I have implemented?

Like I said, this is already working for me in production and is
transparent to the user. The implementation can be changed at any point
since it is not part of the public API.


Saludos,
Carlos R. Gonzales
+502-30274143

**Comment by graingert** [comment]
@rhyek can you separate out the fix to use tabs for concurrency into a different PR?

**Comment by rhyek** [comment]
Sure. I think I'll have some time over the weekend.

**Comment by graingert** [comment]
@rhyek I see you didn't have time over the weekend, but don't worry!

I've extracted from your PR the bits required for parallel generation of PDFs here: #61 would be good to get a review/rebase from you!

**Comment by westy92** [comment]
@rhyek I think the project is at a decent position to get this merged in!  Could you please sync this with master?  Thanks!

**Comment by rhyek** [comment]
Sorry, I have 0 interest in this PR. If someone else wants to help you out with a new one with my code that's fine.

**Comment by westy92** [comment]
Thank you for your contributions so far!  I'll take it from here and use this as a starting point.

**Comment by straldev** [comment]
@westy92 What's the status on this PR?

**Comment by gregbown** [comment]
@westy92  Was this ever completed? I have the need to render a screenshot prior to agreement because displaying a final pdf in the browser would be equivalent to free. This would be very convenient and streamline the document process. 
@rhyek I read that you use this feature in production but had issues with crashing related to tabs. Did you resolve this issue?

Is this project active? I hope Puppeteer has not eaten everybody's lunch?

**Comment by vivex** [comment]
any update on this?

**Comment by westy92** [comment]
Screenshot functionality has been published in version 0.8.0! 🎉

**Pull request closed without merging**